File: examples/python/api_client.py
"""
IntegrateWise OS API Client Example (Python)

Demonstrates API usage for tasks, notes, conversations, plans,
and agent invocation.
"""

import os
import requests
from typing import Optional, Dict, Any, List


class IntegrateWiseClient:
    def __init__(self, api_key: str, base_url: str = 'https://api.integratewise.co/v1'):
        self.api_key = api_key
        self.base_url = base_url
        self.session = requests.Session()
        self.session.headers.update({
            'Authorization': f'Bearer {api_key}',
            'Content-Type': 'application/json'
        })

    def _request(self, method: str, path: str, data: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        url = f'{self.base_url}{path}'
        response = self.session.request(method, url, json=data)
        response.raise_for_status()
        return response.json()

    # Tasks API
    def list_tasks(self, limit: Optional[int] = None, cursor: Optional[str] = None,
                   order: Optional[str] = None, filters: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        params = {}
        if limit:
            params['limit'] = limit
        if cursor:
            params['cursor'] = cursor
        if order:
            params['order'] = order
        
        # Build filter query params
        if filters:
            for field, condition in filters.items():
                for op, value in condition.items():
                    if isinstance(value, list):
                        value = ','.join(value)
                    params[f'filter[{field}][{op}]'] = value

        query_string = '&'.join([f'{k}={v}' for k, v in params.items()])
        return self._request('GET', f'/tasks?{query_string}')

    def get_task(self, task_id: str, include: Optional[List[str]] = None) -> Dict[str, Any]:
        params = {}
        if include:
            params['include'] = ','.join(include)
        
        query_string = '&'.join([f'{k}={v}' for k, v in params.items()])
        path = f'/tasks/{task_id}'
        if query_string:
            path += f'?{query_string}'
        return self._request('GET', path)

    def create_task(self, task: Dict[str, Any]) -> Dict[str, Any]:
        return self._request('POST', '/tasks', {'data': task})

    def update_task(self, task_id: str, updates: Dict[str, Any]) -> Dict[str, Any]:
        return self._request('PATCH', f'/tasks/{task_id}', {'data': updates})

    # Notes API
    def list_notes(self, limit: Optional[int] = None, filters: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        params = {}
        if limit:
            params['limit'] = limit
        
        if filters:
            for field, condition in filters.items():
                for op, value in condition.items():
                    if isinstance(value, list):
                        value = ','.join(value)
                    params[f'filter[{field}][{op}]'] = value

        query_string = '&'.join([f'{k}={v}' for k, v in params.items()])
        return self._request('GET', f'/notes?{query_string}')

    def create_note(self, note: Dict[str, Any]) -> Dict[str, Any]:
        return self._request('POST', '/notes', {'data': note})

    # Agent Invocation API
    def invoke_agent(self, agent_id: str, input_data: Dict[str, Any],
                     options: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        payload = {
            'input': input_data,
            'options': options or {}
        }
        return self._request('POST', f'/agents/{agent_id}/invoke', payload)

    def get_invocation_status(self, invocation_id: str) -> Dict[str, Any]:
        return self._request('GET', f'/agents/invocations/{invocation_id}')


# Usage examples
async def examples():
    client = IntegrateWiseClient(api_key=os.environ.get('INTEGRATEWISE_API_KEY'))

    # List tasks
    tasks = client.list_tasks(
        limit=50,
        filters={
            'status': {'eq': 'in_progress'},
            'priority': {'in': ['high', 'urgent']}
        },
        order='desc'
    )
    print('Tasks:', tasks['data'])

    # Get task with relationships
    task = client.get_task('550e8400-e29b-41d4-a716-446655440000', include=['relationships'])
    print('Task:', task['data'])

    # Create task
    new_task = client.create_task({
        'title': 'New task',
        'status': 'pending',
        'priority': 'medium',
        'source_tool': 'api',
        'external_ref': 'api-ref-123'
    })
    print('Created task:', new_task['data'])

    # Invoke agent
    invocation = client.invoke_agent('planning-agent', {
        'query': 'Create a plan for Q4 product launch',
        'context_entity_ids': [task['data']['id']]
    }, {
        'timeout': 300,
        'approval_required': True
    })
    print('Invocation:', invocation['data'])

    # Check invocation status
    status = client.get_invocation_status(invocation['data']['invocation_id'])
    print('Status:', status['data'])


if __name__ == '__main__':
    import asyncio
    asyncio.run(examples())

File: examples/js/api-client.js
/**
 * IntegrateWise OS API Client Example (JavaScript/TypeScript)
 * 
 * Demonstrates API usage for tasks, notes, conversations, plans,
 * and agent invocation.
 */

class IntegrateWiseClient {
  constructor(config) {
    this.apiKey = config.apiKey;
    this.baseURL = config.baseURL || 'https://api.integratewise.co/v1';
  }

  async request(method, path, data = null) {
    const url = `${this.baseURL}${path}`;
    const options = {
      method,
      headers: {
        'Authorization': `Bearer ${this.apiKey}`,
        'Content-Type': 'application/json'
      }
    };

    if (data) {
      options.body = JSON.stringify(data);
    }

    const response = await fetch(url, options);
    
    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.error?.message || `HTTP ${response.status}`);
    }

    return response.json();
  }

  // Tasks API
  async listTasks(params = {}) {
    const query = new URLSearchParams();
    if (params.limit) query.append('limit', params.limit);
    if (params.cursor) query.append('cursor', params.cursor);
    if (params.order) query.append('order', params.order);
    
    // Build filter query params
    if (params.filter) {
      Object.entries(params.filter).forEach(([field, condition]) => {
        Object.entries(condition).forEach(([op, value]) => {
          query.append(`filter[${field}][${op}]`, Array.isArray(value) ? value.join(',') : value);
        });
      });
    }

    return this.request('GET', `/tasks?${query.toString()}`);
  }

  async getTask(id, options = {}) {
    const query = new URLSearchParams();
    if (options.include) {
      query.append('include', options.include.join(','));
    }
    const path = `/tasks/${id}${query.toString() ? '?' + query.toString() : ''}`;
    return this.request('GET', path);
  }

  async createTask(task) {
    return this.request('POST', '/tasks', { data: task });
  }

  async updateTask(id, updates) {
    return this.request('PATCH', `/tasks/${id}`, { data: updates });
  }

  // Notes API
  async listNotes(params = {}) {
    const query = new URLSearchParams();
    if (params.limit) query.append('limit', params.limit);
    if (params.filter) {
      Object.entries(params.filter).forEach(([field, condition]) => {
        Object.entries(condition).forEach(([op, value]) => {
          query.append(`filter[${field}][${op}]`, Array.isArray(value) ? value.join(',') : value);
        });
      });
    }
    return this.request('GET', `/notes?${query.toString()}`);
  }

  async createNote(note) {
    return this.request('POST', '/notes', { data: note });
  }

  // Agent Invocation API
  async invokeAgent(agentId, input, options = {}) {
    return this.request('POST', `/agents/${agentId}/invoke`, {
      input,
      options
    });
  }

  async getInvocationStatus(invocationId) {
    return this.request('GET', `/agents/invocations/${invocationId}`);
  }
}

// Usage examples
async function examples() {
  const client = new IntegrateWiseClient({
    apiKey: process.env.INTEGRATEWISE_API_KEY
  });

  // List tasks
  const tasks = await client.listTasks({
    limit: 50,
    filter: {
      status: { eq: 'in_progress' },
      priority: { in: ['high', 'urgent'] }
    },
    order: 'desc'
  });
  console.log('Tasks:', tasks.data);

  // Get task with relationships
  const task = await client.getTask('550e8400-e29b-41d4-a716-446655440000', {
    include: ['relationships']
  });
  console.log('Task:', task.data);

  // Create task
  const newTask = await client.createTask({
    title: 'New task',
    status: 'pending',
    priority: 'medium',
    source_tool: 'api',
    external_ref: 'api-ref-123'
  });
  console.log('Created task:', newTask.data);

  // Invoke agent
  const invocation = await client.invokeAgent('planning-agent', {
    query: 'Create a plan for Q4 product launch',
    context_entity_ids: [task.data.id]
  }, {
    timeout: 300,
    approval_required: true
  });
  console.log('Invocation:', invocation.data);

  // Check invocation status
  const status = await client.getInvocationStatus(invocation.data.invocation_id);
  console.log('Status:', status.data);
}

module.exports = { IntegrateWiseClient };
